### PR TITLE
Allow sidebar links to be specified for Rustdoc

### PR DESCRIPTION
**Pull request by syvb**
This should implement everything needed in `rustc` to implement #72475 (although automatically adding links based on `Cargo.toml` would have to be implemented in `cargo-doc`). It allows an arbitrary links to be added to be added to the left sidebar. These are only displayed on the crate root.

Sidebar links can be specified in two ways:
- On the commandline, like `rustdoc src/lib.rs --sidebar-url Homepage=http://rust-lang.org --sidebar-url "Repository URL"="https://example.com/ a"`
- As a `#[doc]` attribute:
```rust
#![doc(
    sidebar_url(
        name = "Repository",
        url = "https://github.com/rust-lang/rust",
    )
)]
```

## Screenshot
Cropped creenshot of a list of links. Three external links are visible on the left sidebar, directly underneath a header that says "Crate lib": "Homepage", "Refdspo", and "Repository".

## Notes
- [Invalid `#[doc]` attributes are, and always have been silently ignored.](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=0b8a3f5d579e9ff1f25899caccd59367)
- Links are sorted Unicode order.
- Later command-line flags override earlier command-line flags.
- `#[doc]` attributes override commnad-line flags in the case of a disagreement.
- Later `#[doc]` attributes override earlier `#[doc]` attributes.
- Currently it is stable. It could be added as unstable, but since it uses both the `#[doc]` attribute and command line config, it's unclear if it should be enabled by a seperate `rustdoc` flag, or by a rust `feature` (I can't find any info on this).

### CODE REVIEW

File: src/librustc_span/symbol.rs [hunk 1]
```diff
@@ -839,6 +840,7 @@ symbols! {
         test_case,
         test_removed_feature,
         test_runner,
+        text,
```
Comment (GuillaumeGomez) [inline src/librustc_span/symbol.rs:843]: Where is this one used?

File: src/librustdoc/lib.rs [hunk 2]
```diff
@@ -411,6 +411,9 @@ fn opts() -> Vec<RustcOptGroup> {
                 "specified the rustc-like binary to use as the test builder",
             )
         }),
+        stable("sidebar-url", |o| {
```
Comment (GuillaumeGomez) [inline src/librustdoc/lib.rs:414]: It shouldn't be stable.

**Comment by GuillaumeGomez** [comment]
First, thanks for opening this PR! Now, a few things to be noted though:

* I'm not really comfortable with the idea of adding links just like this, or at least added in the sidebar. I think a new section on the crate `index.html` file would better to differentiate between the crates' links and the crates' repositories.
* This should be kept as unstable for the moment (so if this isn't a nightly build, you shouldn't allow this) simply because we can't merge something that might change quickly directly as stable.
* I think this should only be done with command-line option (and through cargo), so the extension of `#[doc()]` seems to be too much for me.

What do you think @rust-lang/rustdoc ?

File: src/librustc_span/symbol.rs (hunk 1 above)
Comment (Smittyvb) [reply src/librustc_span/symbol.rs]: Forgot to remove that, thanks!

**Comment by kinnison** [comment]
I think that if it were a section for index.html then you could just put the links in the inner doc of `lib.rs` or `main.rs` so perhaps sidebar is special in some sense.  I think only via commandline args is fine; and thence `Cargo.toml` via `cargo doc` perhaps.

I quite like seeing where it shows up in the examples in the screenshots, but I wonder if this will end up conflicting with things docs.rs might want to inject?

**Comment by Smittyvb** [comment]
I've fixed the merge conflicts and rebased on to master. I've also renamed the flag to `--crate-root-link`. The flag is currently used like:
```sh
rustdoc +local src/lib.rs -Z unstable-options --crate-root-link Repo=https://a.com --crate-root-link "Web site"="https://a.com"
```
Here's what it looks like on smaller screens:
image

**Comment by Dylan-DPC** [comment]
After a discussion with the reviewer, they don't feel that it is a good idea to merge this at this point in time. If you wish you can find a better way we can tackle this problem. Closing this for now. Thanks for taking the time to contribute.

**Comment by GuillaumeGomez** [comment]
Still remains the worry about those newly added links: basically, it's **very** easy to put whatever you want in there. Just imagine some crates adding ads or whatever, that would be pretty bad and I'm not sure we should allow it.

What do you think @rust-lang/docs-rs ?

**Comment by jyn514** [comment]
The main places documentation shows up are
- docs.rs
- locally

If you document it locally, I don't think we should care about where the link goes, if you want to show yourself ads then fine.
If you document on docs.rs, it will use `cargo doc`. Since rustdoc don't allow the link to be overriden, `cargo doc` will pass `--crate-root-link` and passing it in `rustflags` as well will be an error. So also there I don't think we need to worry about ads.

Is there any concern you had about it other than ads? Otherwise I think this should be reopened.

**Comment by jyn514** [comment]
Oh and the third place is on a self-hosted documentation site, but if you've already gone to a random website I think that implies some level of trust (and you can put arbitrary links on your own site anyway).

File: x.py [hunk 3]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python3
```
Comment (jyn514) [inline x.py:1]: Please revert this change. There are systems that do not have python3 installed (https://github.com/rust-lang/rust/issues/71818#issuecomment-623291126). If you have trouble locally you can make a symbolic link from `python3` to `python`.
```suggestion
 #!/usr/bin/env python
```
